src/storage/tracked.rs

```diff
@@ -0,0 +1,413 @@
+    /// See `Change` for more explanations.
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// use specs::{Change, Component, TrackedStorage, World};
+    ///
+    /// #[derive(Clone, PartialEq)] // `PartialEq` is only necessary for maintain
+    /// struct Comp(u8);
+    ///
+    /// impl Component for Comp {
+    ///     // Will use `DenseVecStorage` by default.
+    ///     type Storage = TrackedStorage<Self>;
+    /// }
+    ///
+    /// let mut w = World::new();
+    /// w.register::<Comp>();
+    ///
+    /// let a = w.create_entity().with(Comp(1)).build();
+    ///
+    /// let events: Vec<_> = w.read::<Comp>().change_events_tracked().collect();
+    /// assert_eq!(events, vec![a.id(), Change::Inserted]);
+    /// ```
+    pub fn change_events_tracked(&self) -> ChangeEvents {
+        self.data.inner.change_events()
+    }
+}
+
+impl<'e, S, T, D> Storage<'e, T, D>
+where
+    S: UnprotectedStorage<T> + Send + Sync + 'static,
+    T: Component<Storage = TrackedStorage<T, S>> + Clone + Send + Sync,
+    D: DerefMut<Target = MaskedStorage<T>>,
+{
+    /// Maintains the `TrackedStorage`.
+    ///
+    /// You can only call this in case your component implements `PartialEq`.
+    /// This will compare the cache with the current storage and generate change events
+    /// in case the `PartialEq` implementation says that two components are different.
+    ///
+    /// If you don't care about `Change::Modified` events, you don't have to call this method.
+    ///
+    /// ## When should I call this method?
+    ///
+    /// You should make sure that it gets called before you need the information which component
+    /// has been modified. E.g. in case you have several systems writing to the component, then
+    /// several reading from it, you can just call `maintain_tracked` once in between.
+    pub fn maintain_tracked(&mut self)
+    where
+        T: PartialEq,
+    {
+        let (set, inner) = self.data.open_mut();
+        unsafe {
+            inner.maintain(set);
+        }
+    }
+
+    /// Resets the tracked storage. This clears all change events.
+    /// You most likely want to do this at the end of every frame.
+    pub fn reset_tracked(&mut self) {
+        let (set, inner) = self.data.open_mut();
+        unsafe {
+            inner.reset(set);
+        }
+    }
+}
+
+/// Storage which tracks changes.
+#[derive(Derivative)]
+#[derivative(Default(bound = "S: Default"))]
+pub struct TrackedStorage<C, S = DenseVecStorage<C>> {
+    changes: Vec<Change>,
+    _marker: PhantomData<C>,
+    old: S,
+    storage: S,
+}
+
+impl<C, S> TrackedStorage<C, S>
+where
+    C: Clone,
+    S: UnprotectedStorage<C>,
+{
+    /// Returns an iterator for all change events.
+    pub fn change_events<'a>(&'a self) -> ChangeEvents<'a> {
+        let inner = self.changes.iter().enumerate();
+
+        ChangeEvents { inner }
+    }
+
+    /// Resets the `TrackedStorage`.
+    pub unsafe fn reset(&mut self, set: &BitSet) {
```

Comment:
why is it unsafe?

Reply:
If the `set` is wrong, this would access uninitialized memory. This is an `UnprotectedStorage`, I'm glad there are safe methods at all :smi